### PR TITLE
SNOW-1619272: fix explode outer parameter type casting causing dynamic table creation failure

### PR DESCRIPTION
**Pull request by sfc-gh-aling**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1619272

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

backend function FLATTEN_CONSTRUCT doesn't accept `False::BOOLEAN` but accept `False`.
to bypass the issue, we don't cast type if the input value is of boolean type on the client side

### CODE REVIEW

**Review by sfc-gh-jdu — APPROVED** [review]

**Review by sfc-gh-aalam — APPROVED** [review]
Approving since we also use the same cast when analyzing `FunctionExpression`

File: src/snowflake/snowpark/_internal/analyzer/analyzer.py [hunk 1]
```diff
@@ -605,7 +605,7 @@ def table_function_expression_extractor(
             sql = named_arguments_function(
                 expr.func_name,
                 {
-                    key: self.analyze(
+                    key: self.to_sql_try_avoid_cast(
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/_internal/analyzer/analyzer.py:608]: to_sql_try_avoid_cast is for numeric types. Does it also work for boolean values?

Comment (sfc-gh-aling) [reply src/snowflake/snowpark/_internal/analyzer/analyzer.py:608]: I added the support for boolean type in the to_sql_try_avoid_cast function

**Review by sfc-gh-evandenberg — APPROVED** [review]
Minor comment otherwise LGTM

File: tests/integ/test_dataframe.py [hunk 2]
```diff
@@ -3015,6 +3015,31 @@ def test_create_dynamic_table(session, table_name_1, is_transient):
...
+        )
+        df.write.mode("overwrite").save_as_table(temp_table)
+        df = session.table(temp_table)
+        df1 = df.select(df.idx, explode(df.lists))
+        df1.create_or_replace_dynamic_table(
+            dt_name, warehouse=session.get_current_warehouse(), lag="1 min"
+        )
+        session.sql(f"alter dynamic table {dt_name} refresh").collect()
+        res = session.sql(f"show dynamic tables like '{dt_name}'").collect()
+        assert len(res) == 1
```
[the remark this answers is not in the document]
Comment (sfc-gh-aling) [reply tests/integ/test_dataframe.py:3038]: the result is composed of dynamic table meta information which I think we don't need to validate in this case.
we care whether the table creation succeeded so checking the length of res is enough

**Comment by sfc-gh-aling** [comment]
daily precommit running: https://github.com/snowflakedb/snowpark-python/actions/runs/10620281367

**Pull request merged**